File: Python/面试经典150题/数组、目录/45_跳跃游戏Ⅱ.py
from typing import List
class Solution:
    def jump(self, nums: List[int]) -> int:
        index, ans = 0, 0
        while index < len(nums)-1:
            if nums[index] == 0:
                return False
            elif index+nums[index] >= len(nums)-1:
                return ans+1
            far = 0
            for i in range(nums[index]):
                if index+i+1 > len(nums)-1:
                    break
                step = nums[index+i+1] + i
                if nums[index+i+1] + i + 1 > far:
                    far =  step
                    next_index = index+i+1
            index = next_index
            ans += 1
        return ans

nums = [2,3,1,1,4]
print(Solution().jump(nums))

File: JS/贪心/45_跳跃游戏Ⅱ.js
var jump = function(nums) {
    if(nums.length === 1) return 0
    function isFinish(index, maxStep, step){
        if(index >= nums.length-1 || nums[index]+index >= nums.length-1) return step+1
        maxStep = 0
        let newIndex = index
        for(let j=1; j<=nums[index]; j++){
            if(nums[index+j]+j >= maxStep) {
                newIndex = index+j
                maxStep = nums[index+j] + j
            }
        }
        maxStep = maxStep - (newIndex - index)
        return isFinish(newIndex, maxStep, step+1)
    }
    return isFinish(0, nums[0], 0)
};
console.log(jump([2,3,1,1,4]))